hathor/manager.py

```diff
@@ -507,6 +510,128 @@ def _initialize_components(self) -> None:
+            self.log.warn('The last started time is to the future of the current time',
+                          started_at=started_at, last_started_at=last_started_at)
+
+        # TODO: this could be either refactored into a migration or at least into it's own method
+        # After introducing soft voided transactions we need to guarantee the full node is not using
+        # a database that already has the soft voided transaction before marking them in the metadata
+        # Any new sync from the beginning should work fine or starting with the latest snapshot
+        # that already has the soft voided transactions marked
+        for soft_voided_id in settings.SOFT_VOIDED_TX_IDS:
+            try:
+                soft_voided_tx = self.tx_storage.get_transaction(soft_voided_id)
+            except TransactionDoesNotExist:
+                # This database does not have this tx that should be soft voided
+                # so it's fine, we will mark it as soft voided when we get it through sync
+                pass
+            else:
+                soft_voided_meta = soft_voided_tx.get_metadata()
+                voided_set = soft_voided_meta.voided_by or set()
+                # If the tx is not marked as soft voided, then we can't continue the initialization
+                if settings.SOFT_VOIDED_ID not in voided_set:
+                    self.log.error(
+                        'Error initializing node. Your database is not compatible with the current version of the'
+                        ' full node. You must use the latest available snapshot or sync from the beginning.'
+                    )
+                    sys.exit(-1)
+
+                assert {soft_voided_id, settings.SOFT_VOIDED_ID}.issubset(voided_set)
+
+        # TODO: move support for full-verification here, currently we rely on the original _initialize_components
+        #       method for full-verification to work, if we implement it here we'll reduce a lot of duplicate and
+        #       complex code
+        self.tx_storage.indexes._manually_initialize(self.tx_storage)
+
+        # Verify if all checkpoints that exist in the database are correct
+        try:
+            self._verify_checkpoints()
+        except InitializationError:
+            self.log.exception('Initialization error when checking checkpoints, cannot continue.')
+            sys.exit()
+
+        # restart all validations possible
+        self._resume_validations()
+
+        # XXX: last step before actually starting is updating the last started at timestamps
+        self.tx_storage.update_last_started_at(started_at)
+        self.state = self.NodeState.READY
+        self.log.info('ready')
+
+    def _verify_checkpoints(self) -> None:
+        """ Method to verify if all checkpoints that exist in the database have the correct hash and are winners.
+
+        This method needs the essential indexes to be already initialized.
```

Comment:
Which ones are the essential ones?

Reply:
All the non-optional, what should I call them? I tried to avoid enumerating all of them. Or rather, it might only _actually_ need a subset of them, but we don't have the notion of splitting them, I mostly meant that this has to be called after the indexes were initialized, but we don't care about the optional indexes.

Reply:
Got it. That's ok. We could be more explicit here but it's ok to leave to the reader the task to dive deep into the index dependencies.